### ISSUE TITLE
Inaccurate results with operations between Floats of different precision

### ISSUE THREAD

**Issue by weakit** [sympy/sympy#20071]
```python
>>> a = N(pi)
3.14159265358979
>>> b = Float('2.14159', 6)
2.14159
>>> a - b
1.00000253518159
```

Shouldn't this output to `1.00000265358979` instead?

If I understand correctly this is due to how mpmath works/is used internally. It should be possible to make it so that only the precise bits are opertaed with, leaving everything else, right?

---

Also a suggestion, I don't know if this may help, but can you consider allocating a few bits for extra precision?

To make myself clear what I'm suggesting is to consider a float to be precise to it's precision minus say 20-30 bits. This might result in better precision overall, and wouldn't really be hard to implement.
 
Say you have the normal 53-bit float, it'll be stored in memory with 53-bits and all operations will also be carried out at the same size. But only the precise bits are displayed, so the remainng bits are rounded off. (only for display, it'll still be stored/used at the same size, here 53 bits)

I feel like this would result in much better precision and accuracy.

For example currently:

```python
>>> Float('456.457') - Float('456')
0.456999999999994
``` 

but with extra precision, and then rounding off to the "precise" digits after the calculation you would get the correct answer: `0.4570000000`

**Comment by oscarbenjamin** [comment]
> Shouldn't this output to `1.00000265358979` instead?

The result will only be accurate to the given precision of the inputs. Note that binary digits are used for a Float internally so `Float('2.14159', 6)` will not be exactly equal to `2.14159`:
```
In [12]: Rational(Float('2.14159', 6))                                                                                                         
Out[12]: 
561405
──────
262144

In [13]: Rational(Float('2.14159', 6)).n()                                                                                                     
Out[13]: 2.14159011840820

In [14]: Rational(Float('2.14159', 6)).n() - Rational('2.14159')                                                                               
Out[14]: 1.18408203242382e-7
```

**Comment by weakit** [comment]
I understand that internally Floats use mpmath numbers. But shouldn't the lower precision Float in my example be converted to `2.14159000000000` before operation to maintain accuracy?
Either that, or return a lower precision Float.
I see that this is more of a preference, so I would like your comments on the matter.
Speaking about comments, any comments on my suggestion? @oscarbenjamin

**Comment by oscarbenjamin** [comment]
> shouldn't the lower precision Float in my example be converted to `2.14159000000000`

You've already created it with a precision of 6 decimal digits (really 23 binary digits) so it no longer represents that number exactly. Why would it be converted to 2.14159000000000 when its actual value is 2.14159011840820?

I agree that the subtraction could give a lower precision Float though.

I'm not entirely sure I understand what thi

> To make myself clear what I'm suggesting is to consider a float to be precise to it's precision minus say 20-30 bits.

I'm not sure I understand what this means.

You need to be careful in explaining this and also careful to define what "precision" means as there are two different sense of the word. Normally in a multiprecision floating point library such as mpmath "precision" essentially refers to the number of digits used in the representation. You can see that in mpf form of the float:
```julia
In [50]: Float('2.14159', 6)._mpf_                                                                                                             
Out[50]: (0, mpz(561405), -18, 20)

In [51]: 561405 * 2**-18                                                                                                                       
Out[51]: 2.141590118408203
```
The 20 in mpf form is the "bit count":
http://mpmath.org/doc/current/technical.html

Sympy wraps mpmath's mpf in a Float which then has its own separate notion of precision:
```julia
In [54]: Float('2.14159', 6)._prec                                                                                                             
Out[54]: 23
```

Sympy's `_prec` is used to keep track of the accuracy in `evalf`:
```julia
In [68]: f = Integral(x, (x, 0, 1)).evalf()                                                                                                    

In [69]: f                                                                                                                                     
Out[69]: 0.500000000000000

In [70]: f._mpf_                                                                                                                               
Out[70]: (0, mpz(1), -1, 1)

In [71]: f._prec                                                                                                                               
Out[71]: 53
```

**Comment by weakit** [comment]
Sorry, I should've written it better. I read it again and I see why I couldn't get the idea across.
I knew it would be hard for me to explain, but I'm going to try to give it another shot (with bullet points!).

Also, I'm going to be using decimal digits for simplicity instead of bits to measure precision and accuracy, but the idea is the same and works with binary bits.

This is what I'm proposing:

- Each float will have a new metric called _accuracy_, which will be a bit lesser than the value of its precision. 
  Whenever I say precision, I'm talking about the precision tracked by sympy, `_prec`.

- The difference between a float's precision and its accuracy is what I'm going to call _extra precision_. 

- A 10-digit accurate float, with 5-digits of extra precision, will be the same as a normal 15-digit precise float.
  This float will function in every way exactly the same as a 15-digit precise float currently functions right now in sympy. All operations will occur at its specified (here, 15-digit) precision.

- The only difference being that only the accurate digits of the float will be displayed. By displayed, I mean being printed, the same way how `Float('2.14159', 6)` is shown only as `2.14159` instead of what it actually represents internally, `2.141590118408203...`.

- `evalf` will return floats accurate up to the requested accuracy, not precision.
  `N(pi, 20)` will return a 20-digit accurate float, _precise_ to (20 + extra precision) digits.

Why I'm proposing this:

- It helps to maintain accuracy while still using binary floating-point arithmetic. (It wouldn't require significantly more resources to run, compared to using decimal floats) 

- It mitigates simple errors that shouldn't happen:
  ```python
  >>> Float('456.457') - Float('456')
  0.456999999999994
  ```
  With the new implementation, if the default float is considered accurate up to 10-digits, with the same 15-digits of precision, the answer will still result in `0.456999999999994` but since it's considered accurate only to 10-digits, it will be _displayed_ as `0.4570000000` which is what you get when you round off the number to 10-digits. 
Note that the float is still `0.456999999999994` and that is what will be used in further operations. It is only being _displayed_ as `0.4570000000`.

- I don't think this wouldn't be very hard to implement. 

I feel being more accurate is worth the few extra bits in memory or the slight loss in precision.

I'd appreciate it if someone could comment and critique on this again, as I'm still learning the ropes with software dev.
@oscarbenjamin?

**Comment by oscarbenjamin** [comment]
I think that your notion of "accuracy" is what the `_prec` attribute is *supposed* to represent. Note that using the word accuracy as distinct from precision is potentially more confusing because Mathematica uses these two words for a different meaning:
https://reference.wolfram.com/language/guide/PrecisionAndAccuracyControl.html

It sounds to me that you are saying that the mpf should internally have a few more bits than the `_prec` of the Float. That maybe seems reasonable. What we don't really have now is a clear (documented) definition of what any of these things actually mean in sympy so just making the definition clear would be good I think.

There is also more broadly the question of interpreting what accuracy should mean in subsequent calculations though. As you say maybe `Float('456.457') - Float('456')` should not give a `Float` with 53-bit precision.

The `.evalf` routines (in `sympy/core/evalf.py`) internally keep track of "precision" separately from the mpf. At the end they will combine the precision with the mpf in a single `Float` such that the precision should represent the error from the calculation (I'm not sure how consistent this is but that's the idea). In order to evaluate something like a subtraction `.evalf` will increase the number of digits used for the two terms being subtracted until the subtracted result has the expected precision.

I guess that you are suggesting that something similar should happen with `a - b` where `a` and `b` are `Float`. In that case of course it isn't possible to increase the precision of `a` and `b` but it could be possible to return a `Float` with reduced precision when cancellation has increased the relative "error" in the result.

**Comment by weakit** [comment]
> What we don't really have now is a clear (documented) definition of what any of these things actually mean in sympy so just making the definition clear would be good I think.

Looking up an online dictionary gives me this:

Accuracy is defined as:
> the degree to which the result of a measurement, calculation, or specification conforms to the correct value or a standard.

Whereas precision is defined as:
> refinement in measurement, calculation, or specification, especially as represented by the number of digits given.

Wolfram pretty much defines precision as the number of significant digits in a given number, 
and accuracy as the significant number of digits _right of the decimal point_.

This _is_ confusing, and I don't see how this would be helpful.

If you'd ask me to define the "accuracy" and "precision" of a number right off the top of my head, this is probably what I'd tell you:

- Accuracy would be the number of significant digits _guaranteed to be correct_, starting from the left of the number.
- Precision would be the number of significant digits in a number.

I think this is what most people would think too. 

As an example, a pair of callipers: 
Its precision would be the difference in value between the smallest divisions on its scale (or its least count), which works out to the number of significant digits in its measurement; its accuracy would be the degree to which the measurement is correct, which pretty much is the number of _correct_ significant digits.

Another example: `3.1482` would be Pi _accurate_ to 3 digits, and _precise_ to 5 digits.

I understand that the definition is a bit loose, but I feel like this would be a good starting point.

Whenever I'll be using the words precision and accuracy, this is what I'll be referring to.

---

> It sounds to me that you are saying that the mpf should internally have a few more bits than the _prec of the Float. That maybe seems reasonable.

Yes, somewhat. Similar to how Wolfram's MachinePrecision/ArbitraryPrecision numbers behave:

![image](https://user-images.githubusercontent.com/34541656/93057817-2253e800-f68c-11ea-9d89-352ade35c5ef.png)

![image](https://user-images.githubusercontent.com/34541656/93047200-91741100-f679-11ea-9c22-99943166698c.png)

With the definitions of accuracy and precision that I set previously, the accuracy of the output is the number of digits normally displayed, and the precision would be the number of digits displayed in the FullForm output. 

The difference between this precision and accuracy is what I defined as _extra precision_ in my previous comment, for the lack of better words.

`_prec` currently tracks precision, and is supposed to track accuracy as well, but does not guarantee it (I'm not sure how that is supposed to work).

Instead of the mpf having a few more bits than `_prec`, I'm suggesting the use of a new attribute to track accuracy, which is lesser than `_prec`. The `_prec` will still have the same bits as the internal mpf, but will be greater than the new "accuracy" of the float. 

Calling `Float(something, x)` or `something.evalf(x)` should return an approximation _accurate_ to `x` digits, but precise to a higher value `x + p` digits. This doesn't mean `evalf` should calculate the approximation accurately only up to `x` digits. Instead, calculate up to  `(x + p)` digits and return that as a Float that is _considered_ accurate up to `x` digits, with a precision of `(x + p)` digits, even though it is accurate up to its complete precision. By "considered accurate up to `x` digits" I mean having this new accuracy attribute set to `x`.

Always try to return a result as accurate as possible, but only _guarantee_ accuracy up to how much is asked.

Floats should be printed up to their accurate digits defined by this new attribute, instead of up to `_prec` like before.

This value of extra precision `p` is still something that should be decided, along with whether it should be constant for all floats, or should be dependent on the accuracy.

This might also help with guaranteeing numerical approximation, issue #19940.

---

> I guess that you are suggesting that something similar should happen with `a - b` where `a` and `b` are `Float`. In that case of course it isn't possible to increase the precision of `a` and `b` but it could be possible to return a `Float` with reduced precision when cancellation has increased the relative "error" in the result.

That won't really be necessary because if you do implement this "extra precision", it should just work:

![image](https://user-images.githubusercontent.com/34541656/93051545-4b6f7b00-f682-11ea-8137-eeb988fb3dd7.png)

---

I do understand that this is a lot, and might break a few things, but I don't see why this can't be implemented. I think I can work on this, given enough time and a lot of help (I'm still new); although this is still a decision to make. 
@oscarbenjamin do you mind commenting on this again?

**Comment by oscarbenjamin** [comment]
> I don't see how this would be helpful.

The Wolfram definitions of accuracy and precision are related to the idea of absolute and relative error. The issue is that it is hard to give a notion of precision to something that is possibly zero e.g.:
```julia
In [10]: expr = sin(1)**2 + cos(1)**2 - 1                                                                                                      

In [11]: expr.evalf()                                                                                                                          
Out[11]: -0.e-124

In [12]: expr.evalf()._prec                                                                                                                    
Out[12]: 1
```
In that context it can be more useful to think in terms of absolute rather than relative error.

We don't need to use the words accuracy and precision in the same way as Wolfram but I can see the usefulness of their notion of accuracy.

> Instead of the mpf having a few more bits than `_prec`, I'm suggesting the use of a new attribute to track accuracy, which is lesser than `_prec`. The `_prec` will still have the same bits as the internal mpf, but will be greater than the new "accuracy" of the float.

I think this is what the `_prec` attribute is *already* supposed to do. That's why it is distinct from the precision as recorded in the `_mpf_`. It seems that in some way the behaviour of `_prec` is not what you are expecting so that could be changed but I don't see why an additional attribute would be needed.

Is the difference with mathematica not just the way that the `Float` is *printed* rather than the actual value that it has or the associated precision?

**Comment by weakit** [comment]
> In that context it can be more useful to think in terms of absolute rather than relative error.

Thanks for that, I guess it could be useful.

> I think this is what the `_prec` attribute is _already_ supposed to do. That's why it is distinct from the precision as recorded in the `_mpf_`. It seems that in some way the behaviour of `_prec` is not what you are expecting so that could be changed but I don't see why an additional attribute would be needed.

That works too. But if a Float is going to have a different "precision" and "accuracy", I just thought having an additional attribute would be convenient, and less confusing.

Then right now what I would like is to have the `_mpf_` precision of a Float to be higher than its `_prec`.
~~That would also probably give similar results.~~
It works as expected. After a few dirty edits, and having the `_prec` to be lower than the `_mpf_` by 20 bits this is what I got:

```python
>>> from sympy import *
>>> x = Float('456.457')
>>> x._prec
33
>>> x._mpf_
(0, 4015038232637997, -43, 52)
>>> x - 456  # previously this would give 0.456999999999994
0.457000000
```

But then there's the question of how many more bits you would use to keep results accurate, and whether it should even be a constant for all Floats. Mathematica uses anywhere from 1-2 to as many as around 20 digits (that's the most I've seen, this isn't something you would really check normally) of extra precision depending on the precision of the number. Note that I said digits, not bits.

Also one more thing:

> but it could be possible to return a Float with reduced precision when cancellation has increased the relative "error" in the result.

I think mpmath already handles this on its own because I noticed that the mpmath precision is sometimes lower than sympy's `_prec`:

```python
>>> x = Float('456.457')  # this is without the edits done previously
>>> x._prec, x._mpf_
(53, (0, 4015038232637997, -43, 52))
>>> x - 456
0.456999999999994
>>> _._prec, _._mpf_  # mpf prec is at 42, as opposed to sympy's _prec which is still at 53
(53, (0, 4019814511149, -43, 42))
>>> 4019814511149 * 2 ** -43  # result at default 53 bits
0.45699999999999363
>>> round(_, 12)  # rounding off to 42 bits gives the correct answer
0.457
``` 


> Is the difference with mathematica not just the way that the Float is printed rather than the actual value that it has or the associated precision?

I don't seem to understand what you're asking.
Could you rephrase that?

**Comment by oscarbenjamin** [comment]
> Mathematica uses anywhere from 1-2 to as many as around 20 digits

Can you explain how this actually works in Mathematica?

I'm not sure I understand why there would be a variable number of additional digits.

> > Is the difference with mathematica not just the way that the Float is printed rather than the actual value that it has or the associated precision?
> 
> I don't seem to understand what you're asking.
> Could you rephrase that?

You seem to want the result to display as 0.457 although you know that the numerical value is closer to 0.45699999999999363. To me that suggests that the actual issue is not the exact numerical value of the Float but the way that it is displayed as a string.

**Comment by weakit** [comment]
> Mathematica uses anywhere from 1-2 to as many as around 20 digits
I tried to look at the documentation to explain that behaviour, but I couldn't really find anything.

I did a little experimenting and figured that it's mostly for performance benefits:
![image](https://user-images.githubusercontent.com/34541656/93455386-eaea7300-f8f9-11ea-819b-debe3ccd5cbc.png)

There might be other reasons, but I'm not sure as to what that can be.

One thing I did find tho, is that Mathematica tries to maintain 9-10 digits of extra precision for any calculation.

This doesn't happen however with a few cases, for example when input precision is specified explicitly:
![image](https://user-images.githubusercontent.com/34541656/93455908-97c4f000-f8fa-11ea-88ba-f1eab12d46f7.png)
![image](https://user-images.githubusercontent.com/34541656/93456352-32253380-f8fb-11ea-9ef4-62727514bf53.png)

Another thing I noticed is that numbers change precision based on operations:
![image](https://user-images.githubusercontent.com/34541656/93458436-ecb63580-f8fd-11ea-89c8-82b7a0aa517b.png)

That the internal `_mpf_` precision of a sympy Float also changes similarly:
```python
>>> p = lambda f: (f._prec, f._mpf_[-1])  # shows sympy, mpf precision

>>> a = Float('456.457')
>>> p(a)
(53, 52)

>>> a = a - 456
>>> p(a)  # mpf prec drops
(53, 42)

>>> a = a + 456
>>> p(a)  # mpf prec rises
(53, 52)
```

I don't think this would still work if you just set sympy's _prec to ` _mpf_` precision, and it'll probably create a lot of other problems as well. I do find this noteworthy, however.

> You seem to want the result to display as 0.457 although you know that the numerical value is closer to 0.45699999999999363. To me that suggests that the actual issue is not the exact numerical value of the Float but the way that it is displayed as a string.

Not really. What I was asking (and am still asking for) is to have the internal value be 0.4569999999999999999999999999363 (a bit more closer to the actual value, ie higher internal precision) instead of 0.45699999999999363. That would result in it being displayed as 0.457 correctly, instead of the 456999999999994 that you currently get.

**Comment by oscarbenjamin** [comment]
It looks like mathematica's internal representation for the mantissa uses 64-bit digits.

> Another thing I noticed is that numbers change precision based on operations:

Yes and that makes sense if you think in terms of absolute error (what Mathematica calls accuracy):
```
(456.457 +- 1e-12)  -  (456 +- 1e-12)  =  (0.457 +- 2e-12)
```
The errors are all order 1e-12 but considered as a fraction of the larger input numbers it gives a higher "precision" then for the smaller result of the subtraction.

> What I was asking (and am still asking for)

It seems to me that there are two separate suggestions here:

1. Have the precision of a Float correctly updated in arithmetic operations that should give lower precision results.

2. Use a higher internal precision than would be implied by the `_prec` attribute.

You are welcome to try these things out. Obviously the reason that we want to keep the internal precision as low as possible is performance so we would need to see how that is impacted by any change like that.

I would also like any behaviour for `Float` and `evalf` to be clearly and unambiguously documented somewhere so that the expectations and meaning of precision etc can be understood by both users and contributors.

**Comment by weakit** [comment]
Sure I'll give it a shot.
